src/tensor.hh

```diff
@@ -0,0 +1,297 @@
+    * 
+    * if multiple input Tensors are specified, they can be separated with a comma:
+    * input<tensor, tensor>        : 2 non-batched 2D tensor containing float32 values
+}}} */
+
+#ifndef SRC_TENSOR_HH_
+#define SRC_TENSOR_HH_
+
+#include <string>
+#include <map>
+#include <cassert>
+#include <regex>  // NOLINT [build/c++11]
+#include <utility>
+#include <vector>
+#include <iostream>
+#include "type.hh"
+
+#define DEFAULT_TORCH_TYPE "torch::kFloat32"
+#define DEFAULT_CPP_TYPE   "float"
+
+class TensorMode {
+ public:
+  bool batched;             // true if tensor has batch dimension
+  std::string torch_dtype;  // torch datatype of the Tensor
+  std::string cpp_dtype;    // C++ analog of the torch datatype
+  int n_dims;               // number of dimensions
+
+  TensorMode() :
+    batched(false), torch_dtype(DEFAULT_TORCH_TYPE),
+    cpp_dtype(DEFAULT_CPP_TYPE), n_dims(2) {}
+
+  TensorMode(bool batched, std::string &torch_dtype,
+             std::string &cpp_dtype, int n_dims) :
+             batched(batched), torch_dtype(torch_dtype),
+             cpp_dtype(cpp_dtype), n_dims(n_dims) {}
+
+  TensorMode(const TensorMode &other) :
+    batched(other.batched), torch_dtype(other.torch_dtype),
+    cpp_dtype(other.cpp_dtype), n_dims(other.n_dims) {}
+
+  TensorMode(TensorMode &&other) :
+    batched(other.batched), torch_dtype(std::move(other.torch_dtype)),
+    cpp_dtype(std::move(other.cpp_dtype)), n_dims(other.n_dims) {}
+
+  TensorMode& operator=(const TensorMode &other) {
+    batched = other.batched;
+    torch_dtype = other.torch_dtype;
+    cpp_dtype = other.cpp_dtype;
+    n_dims = other.n_dims;
+    return *this;
+  }
+
+  static bool is_tensor(const std::string &input) {
+    return input.find("tensor") != input.npos;
+  }
+
+  /*
+   * Tensor type notation in GAP-L input<> declaration: "["batched"][ND]tensor[dtype]"
+   * allowed dtypes: F32 (default), F64, I32, I64
+   * max number of dimensions (includes batch dimension (if it exists)): 99 (default: 2)
+   * examples:
+   *  - input<batchedtensorF64> : batched 2D tensor containing float64 values
+   *  - input<batched4DtensorI32> : batched 4D tensor containing int32 values
+   *  - input<tensorF64> : non-batched 2D tensor containing float64 values
+   *  - input<batchedtensor> : batched 2D tensor containing float32 values
+   */
+  static TensorMode get_tensor_mode(const std::string &input) {
+    assert(is_tensor(input));
+
+    static hashtable<std::string, std::pair<std::string, std::string>>
+    torch_type = {{"I32", {"torch:kInt32", "int"}},
+                  {"I64", {"torch::kInt64", "long"}},
+                  {"F32", {"torch:kFloat32", "float"}},
+                  {"F64", {"torch::kFloat64", "double"}}};
+
+    // check if input Tensor is batched (default: false)
+    bool batched = input.find("batched") != input.npos;
+
+    // check the number of dims of the input Tensor (default: 2)
+    std::regex regex("[a-zA-Z0-9]*([0-9]{1,2})D[a-zA-Z0-9]*");
```

Comment:
I don't think PyTorch tensors have a dimension limit. But how realistic is it that someone has a tensor with e.g. 90 dimensions? Since I wanted to at least support 10D tensors, this regex needs to parse a 2-digit number, so 99 is the natural hard limit. Would you prefer this to be unlimited instead?